### PR TITLE
Feature/events: Association Project

### PR DESCRIPTION
**Pull request by oxenprogrammer**
## Association Project
- A private tiny eventbrite project where authenticated user creates events and authorized user invites users to events..

## Purpose of the Pull Request
This Pull Request Demonstrates;
- One-to-Many and Many-to-Many through Associations.
- Demonstrate ability to create a simple authentication using sessions.
- Demonstrates the flow of Information in the MVC model.

## Models
- User
- Event
- Invitation

## Controllers
- Users
- Events
- Invitations
- Sessions

## Views
- Users
- Invitations
- Sessions
- Events

## Unit Tests
- Perform Unit Test using inbuilt test and minitest
- Run `rails t` to test the Application

### CODE REVIEW

**Review by NtwaliHeritier — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUESTED🛑🛑🛑🛑**

Hello @oxenprogrammer 
Great work on this project. I am very impressed with how professional the repo looks and how you added stylings.
However, let's make few changes to get it approved.

[REQUIRED]: In your readme file, specify how to run the webpacker install, I had a lot of issues with it and the bootstrap.

[REQUIRED]: When I click on my username in the navbar, It takes me to an unexisting page. Kindly put a user show page or simply make it a span, so that it is unclickable.

[REQUIRED]: The upcoming events part is not showing the events to come. Kindly fix it.
You can add a list to it. You don't need to first click on new event to see them.

[REQUIRED]: For the past event, Kindly make the same changes as requested in the previous change request.

[REQUIRED]: The private events word in the navbar can be a link to the home page. Right now when clicked it is showing the # tag and this is causing issues when trying to navigate elsewhere. This will help with easier navigation as well.

[REQUIRED]: On the users show page, you should see a list of events that the user will be attending.

I am going to show other code improvements to make below.

Make the requested changes and ask for another review👍👍
You can make it👍👌
Happy coding💻💻💻

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  def show
+    @event = Event.find(params[:id])
+    @creator = User.find(@event.creator_id)
+    @event_invitees = Invitation.list_invitees(@event)
+    @invitees = Invitation.name_invitees(@event_invitees)
+  end
+
+  def create
+    @event = Event.new(event_params)
+    @user = User.find(current_user.id)
```
Comment (NtwaliHeritier) [inline app/controllers/events_controller.rb]: This can be simplified by writing @event = current_user.events.build(event_params)

Comment (oxenprogrammer) [reply app/controllers/events_controller.rb]: Thank you @ntwaliHeriter for your review.

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 

## Hey Happy Coder, 👋🏽 
Thanks for the changes made, I applaud you for your effort. We however have a few more changes to make to get this project approved.

### REQUIRED CHANGES ♻️ 

PREVIOUSLY REQUESTED
- [x] Let the Private Events heading in the navbar link to the homepage, instead of the logged-in user page. 


ADDITIONAL CHANGES

- [x] When I (say JinjaBread) create the first event, I can make an invite to myself and even to others like Celine and Cindy. We would then be invited. If I go ahead and create the second event, I can not invite any person that I previously invited for the first event and that is Celine, Cindy and I. This happens for all invites that I make and had already invited before. This is the pop-up that I get.
`Invitation failed, Cindy is already invited`
Obviously, this shouldn't be the expected app behavior since we want different events to have a number of invited persons even when they are already invited to other events, and not a one-person one-event kind of app.
Hope you understand my explanation and you're going to look into it.

- [x] I have successfully invited a person to a past event. This shouldn't be the app behavior. You can remove the invite button for all past events.

- [x] Validate the event's date form field. With this input `11yyyyy11`, I had this creation date instead `2021-03-11 00:00:00 UTC`

### Kindly make all the requested changes and then submit another CRR.

**Cheers 🍹 and Happy coding 💻** 

## Reach me on slack for a chit-chat 🆘  Roy Mukuye

**Comment by oxenprogrammer** [comment]
@mke2111 Thanks for the review, it's really insightful. May you please clarify on this  `Let the Private Events heading in the navbar link to the homepage, instead of the logged-in user page.`. When I am not signed in, my home page (root) is the signup page, and once logged in, my home page is the user page. `What is supposed to be the home page in this case?`

**Comment by mke2111** [comment]
Hi @oxenprogrammer,

What the previous reviewer and I meant was that once a user is logged in clicking on the private events word, which is in the extreme left of the navbar, should redirect to the homepage.

If I am on say the events page and I click on that word, I'm redirected to the user's page. Yet the same functionality still exists. It is better for navigation purposes to have the Private Events redirect to the homepage, instead of the user's page

I hope this is now clear to you.

Happy coding

**Comment by oxenprogrammer** [comment]
@mke2111 The user page is my home page. I'm I supposed to have something different? Also, I had combined the date and time as one field, you do advice that  I separate them into two as to validate as you have suggested? Thank you

**Comment by mke2111** [comment]
If its stated that the homepage should be the user page, then it's okay. But the homepage should have all the events, past and future. Then users page should have all events the user is to attend or has attended.

Best you separate the date and time and then validate each. But if you choose to keep the same format, you can have a placeholder to be followed by users and then validate it as one.

**Review by mahmoud717 — APPROVED** [review]
### STATUS: APPROVED ✔️ ✔️ 

**hi @oxenprogrammer**

**Great job with making all of the requirements. the project is approved, you can go ahead and merge this pull request**

**Pull request merged**